sprints/sustainability/utils.py

```diff
@@ -43,6 +53,78 @@ def generate_month_range(start: str, end: str) -> Generator[Tuple[datetime, date
+
+    It's possible to regenerate long-term cache by specifying `long_term` argument.
+    """
+    # Determine whether we're be using long-term cache or short-term one. Set keys and timeout accordingly.
+    worklogs_key = settings.CACHE_WORKLOGS_KEY_LONG_TERM if long_term else settings.CACHE_WORKLOGS_KEY
+    issues_key = settings.CACHE_ISSUES_KEY_LONG_TERM if long_term else settings.CACHE_ISSUES_KEY
+    timeout = settings.CACHE_WORKLOG_TIMEOUT_LONG_TERM if long_term else settings.CACHE_ISSUES_TIMEOUT_SHOT_TERM
+
+    # Check if worklogs are missing from cache.
+    required_issues: Set[str] = set()
+    worklogs: Dict[str, Dict[str, str]] = _get_cached_dicts_from_keys(
+        (settings.CACHE_WORKLOGS_KEY, settings.CACHE_WORKLOGS_KEY_LONG_TERM)
+    ) if not long_term else {}
+
+    if missing_worklogs := required_worklogs - worklogs.keys():
+        with connect_to_jira() as conn:
+            retrieved_worklogs: List[Worklog] = conn.worklog_list(list(missing_worklogs))  # type: ignore
+            for worklog in retrieved_worklogs:
+                required_issues.add(worklog.issueId)
+
+        # Check if worklogs are missing from cache.
+        issues: Dict[str, Dict[str, str]] = _get_cached_dicts_from_keys(
+            (settings.CACHE_ISSUES_KEY, settings.CACHE_ISSUES_KEY_LONG_TERM)
+        ) if not long_term else {}
+        new_issues: Dict[str, Dict[str, str]] = {}
+
+        if missing_issues := required_issues - issues.keys():
+            with connect_to_jira() as conn:
+                try:
+                    retrieved_issues = conn.search_issues(
+                        f'id in ({",".join(missing_issues)})',
+                        fields='project', maxResults=0
+                    )
+                except JIRAError:
+                    # We can notice this for long-term cache, as Jira has limits for the header size.
+                    retrieved_issues = []
+                    for chunk in chunks(list(missing_issues), 12):
+                        retrieved_issues = conn.search_issues(f'id in ({",".join(chunk)})',
+                                                              fields='project', maxResults=0)
+            new_issues = {issue.id: {'key': issue.key, 'project': issue.fields.project.name}
+                          for issue in retrieved_issues}
+
+        new_worklogs = {worklog.id: issues.get(worklog.issueId, new_issues.get(worklog.issueId))
+                        for worklog in retrieved_worklogs}
+
+        # We retrieve cache second time to avoid race conditions.
+        with cache.lock(settings.CACHE_ISSUES_LOCK):
+            if new_issues:
+                issues = cache.get(issues_key) or {}
+                issues.update(new_issues)
+                cache.set(issues_key, issues, timeout)
+
+            worklogs = cache.get(worklogs_key) or {}
+            worklogs.update(new_worklogs)
+            cache.set(worklogs_key, worklogs, timeout)
+
+    return worklogs
+
+
```

Comment:
I don't like this approach, but that was the only thing I could think of, due to limited API.

Reply:
Yeah, took me a while to follow this code. Not much to do with the limited API and the need to keep the cache in sync.

One idea is checking if [webhooks](https://developer.atlassian.com/server/jira/platform/webhooks/) would be able to push all updates to sprints so it does not need to pull anything. This would make sure the cache is always in sync and greatly reduce the queries and resource usage in Jira.

Reply:
Oh, that's a nice idea! Could you please add it to the forum thread as an idea of improving things?
However I see one nit with this approach:
> To register a webhook, you can use any of the following methods:
> 
>     - Jira administration console.
>     - Jira REST API (note that the user must have the Jira Administrators global permission).